crates/pink/runtime/src/runtime/extension.rs

```diff
@@ -51,25 +56,32 @@ pub fn get_side_effects() -> ExecSideEffects {
                     {
                         match pink_extension::PinkEvent::decode(&mut &data[..]) {
                             Ok(event) => {
-                                pink_events.push((address, event));
+                                pink_events.push((address.clone(), event.clone()));
+                                is_private_event = event.is_private();
                             }
                             Err(_) => {
                                 error!("Contract emitted an invalid pink event");
                             }
                         }
                     } else {
-                        ink_events.push((address, event.topics, data));
+                        ink_events.push((address.clone(), event.topics.clone(), data.clone()));
```

Comment:
Can we know which tx the event belongs to based on the above information? If not, I suggest to add more data fields since it would be nice if we can group events emitted by a single call.

Reply:
Events aleady grouped by single call because each call would emit a events block.
For contract all, we can further add the `nonce` to the corresponding events block, so that it can be connect to the original tx from the Phala chain.

Added.